design-docs/improved-console.md

```diff
@@ -0,0 +1,283 @@
+# Improved Console Output
+
+## Primary Goals
+* Improve perceived performance by making user aware of background work
+* Modern feel
+* User extensibility
+
+### User Goals for Console
+In priority order, the user wants to answer the following questions:
+
+* When will my build finish?
+    * Soon (should I keep watching it) or later ("I'll check back")
+* Is there anything I should know about my build right now? (e.g. tests have failed, warnings)
+* What is Gradle doing right now?
+* Is my build successful?
+* What command did I execute?
+
+## Story: Display parallel work in-progress independently by operation
+Show incomplete ProgressOperations on separate lines, up to a specified maximum number. 
+
+### Implementation
+A `RichConsole` has 
+ - an appendable `TextArea` (mainArea)
+ - a `ProgressLabel` (consisting of a `ProgressBar` and `StatusText`)
+ - a fixed-height, random-access `TextArea` (progressArea)
+ 
+The mainArea is displayed at the top for rendering important messages like warnings
+and errors.
+
+`--quiet`, `--info`, and `--debug` log level flags affect the output visible in the main TextArea,
+However, the default will change so that `LIFECYCLE` logs are no longer displayed.
+ 
+The `ProgressLabel` is intended to give the user a glance-able indicator that tells
+them whether the build will be finished soon or not. 
+See ["Display build progress as a progress bar through colorized output"](TODO) below.
+
+The multiple progressArea lines are rendered results of a fixed list of `AbstractWorker`s. 
+It displays work-in-progress for each `AbstractWorker`. 
+Forked processes submit BuildOperations with information about "what" is running "where".
+
+Plugins can provide progress indication by publishing `BuildOperations`
+
+A `OperationStatusBarFormatter` can be used to customize the intra-line display of each 
+`ProgressOperation` chain (each `ProgressOperation` has a reference to its parent)
+
+Implementation is similar to `ConsoleBackedProgressRenderer` with throttling and a data
+structure for storing recent updates.
+
+> NOTE: We must avoid use of save/restore cursor position unless we switch to a 
+[terminfo](https://en.wikipedia.org/wiki/Terminfo)-based Console handling instead of using ANSI.
+    
+### Test Coverage
+* A terminal size with fewer than max parallel operations shows (rows - 2) operations in parallel
+* Operation status lines are trimmed at (cols - 1)
+* `System.in` and SystemConsole I/O happens on the mainArea, which is unaffected by other areas
```

Comment:
My hope is that we can implement this in such a way as to avoid the existing bug thereby not requiring "separate" work as opposed to a naive implementation that still breaks + a story to fix it.